src/writer.rs

```diff
@@ -55,30 +56,41 @@ where
     let header_file = File::create(header_path)?;
     if header.vox_offset > 0.0 {
         if is_gz {
-            let mut e = GzEncoder::new(header_file, compression_level);
-            write_header(&mut e, &header)?;
-            write_data(&mut e, &header, data)?;
-            let _ = e.finish()?;
+            let mut writer = ByteOrdered::runtime(
+                GzEncoder::new(header_file, compression_level),
+                header.endianness,
+            );
+            write_header(writer.as_mut(), &header)?;
+            write_data(writer.as_mut(), &header, data)?;
+            let _ = writer.into_inner().finish()?;
         } else {
-            let mut header_writer = BufWriter::new(header_file);
-            write_header(&mut header_writer, &header)?;
-            write_data(&mut header_writer, &header, data)?;
+            let mut writer = ByteOrdered::runtime(BufWriter::new(header_file), header.endianness);
+            write_header(writer.as_mut(), &header)?;
+            write_data(writer, &header, data)?;
```

Comment:
Right. `as_mut()` is a pattern in `byteordered` for mutably borrowing the inner writer in a `ByteOrdered`, so it turns a `&mut ByteOrdered<T, E>` into a `ByteOrdered<&mut T, E>`. A similar case in the wild is in `Option` itself ([`Option::as_mut()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.as_mut)). In this code, it's an alternative to making functions receive a `&mut ByteOrdered<T, E>`. When it's the final writing operation, I can just move the whole writer, so `as_mut()` is no longer needed. It's also called `as_mut()` instead of `as_ref()` for consistency with the rest of the ecosystem.

If there is something that would help make this pattern more readable, I am more than happy to hear more about these concerns and suggestions.

Reply:
Ok, thank you for the explanation. I tried removing some of the `<A, B, S, D, W, E>` because it looks so wrong! :) Sadly, I can't say I have anything useful to add.

Reply:
I made a [joke](https://users.rust-lang.org/t/twir-quote-of-the-week/328/596?u=enet4) about this too. I think that any attempt of dissolving this large list of type parameters is likely to result in unwanted code repetition. Naturally, I would love to be proven wrong here.